### PR TITLE
Fix pyarrow version

### PR DESCRIPTION
**Pull request by pcmoritz**
This fixes both problems found in https://github.com/ray-project/ray/issues/3744

### CODE REVIEW

File: cmake/Modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -118,6 +118,7 @@ if ("${CMAKE_RAY_LANG_PYTHON}" STREQUAL "YES")
 
     # PYARROW_PARALLEL= , so it will add -j to pyarrow build
     set(pyarrow_ENV
+      "SETUPTOOLS_SCM_PRETEND_VERSION=0.11.1-RAY"
```
Comment (robertnishihara) [inline cmake/Modules/ThirdpartyToolchain.cmake]: So does this need to be updated every time we update arrow? If so, we should include a prominent comment in https://github.com/ray-project/ray/blob/8674606e2687074b043129f7e743d6522d015c90/cmake/Modules/ArrowExternalProject.cmake#L17

explaining that people need to update this code.

File: python/build-wheel-macos.sh [hunk 2]
```diff
@@ -67,7 +67,7 @@ for ((i=0; i<${#PY_VERSIONS[@]}; ++i)); do
     $PIP_CMD install --upgrade setuptools
     # Install setuptools_scm because otherwise when building the wheel for
     # Python 3.6, we see an error.
-    $PIP_CMD install -q setuptools_scm==2.1.0
+    $PIP_CMD install -q setuptools_scm==3.1.0
```
Comment (robertnishihara) [inline python/build-wheel-macos.sh]: What was the issue?

File: cmake/Modules/BoostExternalProject.cmake [hunk 3]
```diff
@@ -64,6 +64,6 @@ if (RAY_BUILD_BOOST)
     BUILD_IN_SOURCE 1
     BUILD_BYPRODUCTS ${Boost_BUILD_PRODUCTS}
     CONFIGURE_COMMAND ./bootstrap.sh
-    BUILD_COMMAND bash -c "./b2 cxxflags=-fPIC cflags=-fPIC variant=release link=static --with-filesystem --with-system --with-thread --with-atomic --with-chrono --with-date_time --with-regex -j8 install --prefix=${Boost_INSTALL_PREFIX} > /dev/null"
+    BUILD_COMMAND bash -c "./b2 cxxflags=\"-fPIC -D_GLIBCXX_USE_CXX11_ABI=0\" cflags=-fPIC variant=release link=static --with-filesystem --with-system --with-thread --with-atomic --with-chrono --with-date_time --with-regex -j8 install --prefix=${Boost_INSTALL_PREFIX} > /dev/null"
```
Comment (robertnishihara) [inline cmake/Modules/BoostExternalProject.cmake]: Could you say what this does and what the issue was?

File: python/build-wheel-macos.sh (hunk 2 above)
Comment (pcmoritz) [reply python/build-wheel-macos.sh]: On macos python 2.7 it didn't support SETUPTOOLS_SCM_PRETEND_VERSION

File: cmake/Modules/BoostExternalProject.cmake (hunk 3 above)
Comment (pcmoritz) [reply cmake/Modules/BoostExternalProject.cmake]: Yeah, it caused the linker error I posted in https://github.com/ray-project/ray/issues/3744. There was a function in boost regex that used std::string and that needs the old ABI.

File: cmake/Modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (pcmoritz) [reply cmake/Modules/ThirdpartyToolchain.cmake]: fixed

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**